### ISSUE TITLE
[Docs] Steps on how to create a new Backpack v6 theme

### ISSUE THREAD

**Issue by tabacitu** [Laravel-Backpack/docs#417]
For the first time ever... we'll have an _official_ and _supported_ way to turn an HTML template you buy from GetBootstrap, WrapBootstrap or ThemeForest to a Backpack theme ❤️

Our challenge here... is to make it FAST. Let's aim for 5 hours or less 🤞

**Comment by maurohmartinez** [comment]
[WIP]

# Theme structure

So, you bought a theme from GetBootstrap, WrapBootstrap or ThemeForest, and now you are wondering... where to begin?

`resources/views/vendor/backpack/theme-your-name`

## Namespace

## Blade files

## CSS files

## JS files

**Comment by tabacitu** [comment]
Update tutorial:

---

Here are the steps to easily build a **Backpack** theme using a template you got from **GetBootstrap**, **WrapBootstrap** or **ThemeForest**.

### 1 - Create a view folder anywhere in your `resources/views`.
This is the directory where you will place all the files for your theme.

```bash
mkdir resources/views/my-cool-theme
```

### 2 - In your `config/backpack/ui.php`, add that as the primary view namespace:

```php
    'view_namespace' => 'my-cool-theme.',
```

> **(!) Important**
> 1 - Notice the `.` at the end of the namespace, that's important.
> 2 - The namespace must match the name of the folder created from the previous step.

### 3 - Choose and use a fallback theme
A fallback theme is needed in cases when Backpack attempts to load a view that doesn't exist in your theme. It means you don't need to create all the views in order to create a theme... Phew! You can easily rely on your fallback theme and only create the views you need to customize. In order to do so, edit your config file `config/backpack/ui.php` as it follows:

```php
    'view_namespace' => 'my-cool-theme.',
    'view_namespace_fallback' => 'backpack.theme-coreuiv4::', // <--- this line
```

If it's your first time creating a Backpack theme, we recommend you start from our CoreUIv4 theme, and use that as your fallback. It's the simplest modern theme we have. It uses Bootstrap v5, but doesn't have any extra features you'd need to also support (like Tabler does). If you don't already have it installed, you will need to do `composer require backpack/theme-coreuiv4`

### 4 - Create main blade files

If you refresh the page right now, it will show up IDENTICAL to CoreUIv4. Why? Because you're using all the blade files from that theme (through the fallback system). How can you make _your theme_ look like _your theme_? By overriding some of blade files the fallback theme provides. Similar to how child themes work in Wordpress. 

Feel free to look at your fallback theme's views (eg. `vendor/backpack/theme-coreuiv4/resources/views`). If you create a file with the same file in your theme directory (eg. `resources/views/my-cool-theme`), your view will be picked up.

So let's do that. Let's create _most_ of the files you'll need to customize, to provide _your theme_ with _your style_: 
```
- my-cool-theme/
  - layouts/
      app.blade.php
  - widgets/
      ...here you can override widgets with your own!
  - components/
      ...here you can override widgets with your own!
  - assets/
      - css/
         ...here place all css files provided by your original purchased theme
      - js/
         ...here place all js files provided by your original purchased theme
  - inc/
      styles.blade.php
      scripts.blade.php
      header.blade.php
      menu.blade.php
      footer.blade.php
```

Now let's build those files... Let's start with what makes a theme different.

#### `my-cool-theme/inc/theme_styles.blade.php`

This file should hold all custom CSS that your theme needs. For example:

```php
{{-- You can load files directly from CDNs, they will get cached and loaded from local --}}
@basset('https://cdn.jsdelivr.net/npm/bootstrap@5.3.0/dist/css/bootstrap.min.css')

{{-- You can also load files from any place in your application directories --}}
@basset(views_path('my-cool-theme/assets/css/extra.css'))

{{-- You can also write inline CSS blocks --}}
@bassetBlock('my-cool-theme/custom-styling')
<style>
.something {
    border: 1px solid red;
}
</style>
@endBassetBlock
```

Note: Don't forget to load the Bootstrap CSS. Backpack does NOT load it by default.

#### `my-cool-theme/inc/theme_scripts.blade.php`

This file should hold all custom JS that your theme needs:

```php
{{-- You can load files directly from CDNs, they will get cached and loaded from local --}}
@basset('https://cdn.jsdelivr.net/npm/bootstrap@5.3.0/dist/js/bootstrap.bundle.min.js')

{{-- You can also load files from any place in your application directories --}}
@basset(views_path('my-cool-theme/assets/css/extra.js'))

{{-- You can also write inline JS blocks --}}
@bassetBlock('my-cool-theme/custom-scripting')
<script>
 alert('got here');
</script>
@endBassetBlock
```

Note: Don't forget to load the Bootstrap JS. Backpack does NOT load it by default.

#### `my-cool-theme/layouts/default.blade.php`

This will be the primary layout for your theme. So it has to contain a full HTML page: the HTML doctype declaration, head, body components, everything. The following example will help you get started.

```html
<!DOCTYPE html>
<html lang="{{ app()->getLocale() }}" dir="{{ backpack_theme_config('html_direction') }}">
<head>
    @include(backpack_view('inc.head'))
</head>
<body class="{{ backpack_theme_config('classes.body') }}">

    @include(backpack_view('inc.sidebar'))

    <div class="wrapper d-flex flex-column min-vh-100 bg-light">

        @include(backpack_view('inc.main_header'))

        <div class="app-body flex-grow-1 px-2">

            <main class="main">

                @yield('before_breadcrumbs_widgets')
                @includeWhen(isset($breadcrumbs), backpack_view('inc.breadcrumbs'))
                @yield('after_breadcrumbs_widgets')

                @yield('header')

                <div class="container-fluid animated fadeIn">

                    @yield('before_content_widgets')
                    @yield('content')
                    @yield('after_content_widgets')

                </div>

            </main>

        </div>{{-- ./app-body --}}

        <footer class="{{ backpack_theme_config('classes.footer') }}">
            @include(backpack_view('inc.footer'))
        </footer>
    </div>

    @include(backpack_view('inc.bottom'))
</body>
</html>
```

We recommend you copy-paste your own HTML above it, then include the `@directives` where they make sense in your layout. Their names should explain pretty well what they do.

Next up, we'll have to drill down. And move any custom content that's needed for the layout... for example for the sidebar, the header, the topbar... into their own respective views.


#### `my-cool-theme/inc/head.blade.php`

There should be no reason for you to customize this file.

#### `my-cool-theme/inc/sidebar.blade.php`

We recommend you:
- create the file;
- paste the `main_header.blade.php` from your fallback theme;
- paste the HTML content you want;
- copy the useful things from the fallback theme to your own HTML, then delete whatever you don't want;

Do not drill down to customize `sidebar_content.blade.php` too. Menu items work a little different than other views - they are _view components_. So instead of customizing `sidebar_content.blade.php` take a few minutes and customize the menu items HTML, by copy-pasting the `components` directory from your fallback theme, then customizing the files inside it (`menu-item`, `menu-dropdown`, `menu-dropdown-item`, `menu-separator` etc).

#### `my-cool-theme/inc/main_header.blade.php`

We recommend you:
- create the file;
- paste the `main_header.blade.php` from your fallback theme;
- paste the HTML content you want;
- copy the useful things from the fallback theme to your own HTML, then delete whatever you don't want; 

#### `my-cool-theme/inc/breadcrumbs.blade.php`

Most often this will look ok out-of-box, no need to customize. But if you do need to customize the breadcrumbs, follow the same process above to re-use everything you need from the fallback theme file.

#### `my-cool-theme/inc/footer.blade.php`

Most often this will look ok out-of-box, no need to customize. But if you do need to customize the breadcrumbs, follow the same process above to re-use everything you need from the fallback theme file.

#### `my-cool-theme/inc/bottom.blade.php`

There should be no reason for you to customize this file.

**Comment by tabacitu** [comment]
Update: 
- I've added docs in https://github.com/Laravel-Backpack/docs/pull/447
- I've added a theme skeleton in https://github.com/Laravel-Backpack/theme-skeleton

This will provide 80% of the info needed to create a new theme.

HOWEVER. We can still do A LOT more to improve how fast or easy it is to create a theme. We need to create a few themes, going through the tutorial, and improve the tutorial, improve the skeleton, improve the process. The faster we make this, the faster it will be _for us_ to create new themes and _for others_ to do the same.
